### DISCUSSION TITLE
nanoELS H2 for large machine

### DISCUSSION THREAD

**Discussion by shev65** [kachurovskiy/nanoels#112] [Q&A]
Maxim, thanks for developing the device, for
share and help people.
Machine TV-320, the guitar is needed only for threading.
I couldn't find a better project than you!
The screw pitch of the machine is 6 mm. stepper motors must be appropriate.
I will be ordering the following:
https://aliexpi.com/xt7E
https://aliexpi.com/kcN7
Do you think you can take this?

**Comment by kachurovskiy** [comment]
Привет! Спасибо. Можно по-русски если так удобнее :)

Большими станками я ни разу не пользовался поэтому сложно сказать, какое усилие нужно. Твой мотор и драйвер выглядят очень внушительно, мне кажется должно хватить.

В физике не силен но хорошо бы посчитать массу твоей каретки и из 12 Nm получить максимальное ускорение / усилие. [Спросил у народа как это посчитать.](https://physics.stackexchange.com/questions/755025/convert-nm-to-acceleration-for-a-known-mass)

**Comment by shev65** [comment]
Из NEMA34 это самый сильный мотор, китайцы мощнее не предлагают.
Почему я про редукцию и спрашивал. Буду пробовать, думаю получится.
На ютубе, парень показывал гитару на таком моторе, станок поболее моего,
правда не знаю чем закончилось. 
https://www.youtube.com/watch?v=vfHJFxJMpT0&t=1343s

**Comment by shev65** [comment]
Пытался код загрузить в ардуину, вот что пишет:
![2023-04-04_165010](https://user-images.githubusercontent.com/127717539/229846712-d7907b5f-d810-4c19-9d7e-93e9cad8b20a.jpg)

**Comment by shev65** [comment]
как победить?

**Reply by kachurovskiy** [reply]
Зайти в Library Manager и поставить библиотеку FastGPIO 👍

**Comment by shev65** [comment]
Спасибо, помогло!
https://www.youtube.com/watch?v=AwgltElphHg&t=26s
Подобные тесты, со своей гитарой не проводил?

**Reply by kachurovskiy** [reply]
Проводил, сотки не убегают, градусы тоже. https://www.youtube.com/watch?v=uXq3jmSwTEs

**Comment by shev65** [comment]
Отлично!

**Comment by shev65** [comment]
Максим, что там в новой версии?

**Reply by kachurovskiy** [reply]
Привет, какие именно файлы / коммиты ты имеешь в виду? В Н2 в коде я сделал побезопаснее код - теперь совершенно невозможен самозапуск при подаче питания но и On-Off состояние теперь не сохраняется при перезагрузке. Н4 пока в разработке, рано что-то говорить.

**Comment by shev65** [comment]
Привет!
Вчера обновляю страницу, вижу " h4  Remove unused "resetOnStartup" in h2. Initial version of h4."
Думаю, новая версия. Приложенный pdf, как рисунок, в переводчик не загнать, вот и возник вопрос.
У меня соображение есть, пожелания приветствуются?

**Reply by kachurovskiy** [reply]
PDF это просто тех. документация для Н2, там ничего нового, просто описано что сегодня есть.

H4 это код для нового железа с 2-4 осями, само железо пока не готово, код тоже :)

Пожелания приветствуются, конечно! 😀

**Comment by shev65** [comment]
Собрал гитару, погонял на станке, пока без нарезки резьбы.
Для себя,  выявил небольшое неудобство. При нарезании резьбы, во время
отвода суппорта в право, надо удерживать кнопку move right. 
Левая рука постоянно занята, станок большой надо тянуться.
Вижу два пути, или поставить дополнительную кнопку с фиксацией, или кнопку с фиксацией 
сделать программно.
Алгоритм такой: 
Нарезка резьбы включена, правый лимит включен - кнопка с фиксацией.
нарезка резьбы отключена - кнопка без фиксации.
Можно сделать программу универсальной, например закомментировав/разкомментировав
часть кода. Если бы программировал ардуино. сделал бы сам )) но увы.
Интересно твое мнение.

**Reply by kachurovskiy** [reply]
Ясно. Ты не первый кто спрашивает так что видимо придется что-то придумывать :) В режиме ON можно сделать так, что если мы стоим на лимите больше 3 секунд с нажатой кнопкой то кнопка "прилипает" и считается нажатой даже если её отпустить. Чтобы кнопка отлипла нужно нажать и отпустить её. Так подойдет?

**Comment by shev65** [comment]
Да, пойдет! Только в этом случае надо индикацию добавить, чтобы секунды не считать.
Если так сделаешь будет отлично.

**Reply by kachurovskiy** [reply]
Подумал еще, может лучше сделать так что при нажатой кнопке лево/право нажатие на левый/правый лимит делает кнопку "залипшей"? Таким образом не надо ждать, можно сразу нажать две кнопки и пускай себе едет станок на лимит и там стоит. Продолжить движение с лимита нажатием на лево/право.

**Reply by kachurovskiy** [reply]
Или вот ещё вариант - добавить новую настройку 2Т/4Т которую можно переключать одновременным нажатием кнопок лево и право. В режиме 2Т все работает как сейчас, в режиме 4Т клик на кнопку движения запускает движение а второй клик останавливает.

**Reply by shev65** [reply]
Извиняюсь, только увидел сообщение, раньше dithub не пользовался, затупил.
На мой взгляд, одновременное нажатие не совсем удобно.
Мне нравится первый вариант где удерживать 3 секунды.
3 секунды возможно перебор. Но ведь это время можно изменить
в прошивке? Для себя сделаю 05-1 секунда.
В первом варианте ты предлагаешь так:
нажимаем в право, наезжаем на лимит и удерживаем 3 секунды,
кнопка "залипла". я правильно понял?
А вообще ты разработчик, тебе решать ).

**Reply by kachurovskiy** [reply]
Согласен, одновременное нажатие неудобно. Как насчет еще одного значения Step? Сейчас есть `1mm`, `0.1mm`, `0.01mm`, `0.001mm` а добавим еще `limit` при котором простое нажатие (без удерживания) на кнопку лево/право будет отправлять каретку до соответсвующего лимита и на ожидание там. Повторное нажатие прерывает движение или отклеивает каретку от лимита.

**Reply by shev65** [reply]
Это еще удобнее! Каждый может выбрать.
Возможно залипание кому-то и не надо.
Смотри как удобнее. Мне нравится и первый и последний варианты.

**Comment by shev65** [comment]
Максим, почему не реализовал гитару на STM32, там-же ресурсов больше,
наверное было бы проще?
Где удобнее вопросы задавать, здесь или или в телеге?

**Reply by kachurovskiy** [reply]
Как-то начал с Нано, так было проще. А теперь уже больше смысла на ESP32 делать - там частота выше и два ядра - чем я и занят.

**Reply by shev65** [reply]
Какую ESP32 купить для твоего проекта? 
Где-то мелькнуло esp без wifi, по моему где-то у тебя, возможно ошибаюсь.
Смотрю у китайцев, все с wifi, короче муки выбора.  
С wifi конечно тоже надо брать, пригодится.

**Reply by kachurovskiy** [reply]
Новый проект двух-осевой H4 я делаю на ESP32-S3, я думаю большинство дев-бордов подойдет. Другое дело что там используется TCA8418 который считывает нажатия 39 кнопок и сообщает о них всего по 2 пинам. Ну и вдобавок резисторов штук 50, диодов, еще бог знает чего.

Понятно что большинство этих компонент можно в простейшем случае убрать но без TCA8418 я не знаю как обойтись.

**Reply by shev65** [reply]
А смысл без неё обходиться? наоборот хорошее решение, не надо
"лепить" коммутатор на "рассыпухе".

**Reply by shev65** [reply]
(https://www.chipmaker.ru/topic/190144/#comments)
Там есть с расширенной клавиатурой. Куча корпусов.
Начал делать. но как то не пошло. Мне нужна одна ось, пока.
Твой проект, очень вовремя попался. Главное настраивать ни чего не
пришлось. сразу заработал.
